sources/packages/backend/apps/api/src/services/application/application.service.ts

```diff
@@ -1543,6 +1544,90 @@ export class ApplicationService extends RecordDataModelService<Application> {
     });
   }
 
+  /**
+   * Gets all eligible application that can be requested for application
+   * offering change.
+   * @param locationId location id.
+   * @param paginationOptions options to execute the pagination.
+   * @returns list of eligible application that can be requested for
+   * application offering change.
+   */
+  async getEligibleApplicationOfferingChangeRecords(
+    locationId: number,
+    paginationOptions: PaginationOptions,
+  ): Promise<PaginatedResults<Application>> {
+    const applicationQuery = this.repo
+      .createQueryBuilder("application")
+      .select([
+        "application.applicationNumber",
+        "application.id",
+        "currentAssessment.id",
+        "offering.studyStartDate",
+        "offering.studyEndDate",
+        "student.id",
+        "user.firstName",
+        "user.lastName",
+      ])
+      .leftJoin("application.currentAssessment", "currentAssessment")
+      .leftJoin("currentAssessment.offering", "offering")
+      .leftJoin(
+        "application.applicationOfferingChangeRequest",
+        "applicationOfferingChangeRequest",
+      )
+      .innerJoin("application.student", "student")
+      .innerJoin("student.user", "user")
+      .where(
+        new Brackets((qb) =>
+          qb
+            .where("applicationOfferingChangeRequest.id IS NULL")
+            .orWhere(
+              "applicationOfferingChangeRequest.applicationOfferingChangeRequestStatus NOT IN (:...status)",
+              {
+                status: [
+                  ApplicationOfferingChangeRequestStatus.InProgressWithSABC,
+                  ApplicationOfferingChangeRequestStatus.InProgressWithStudent,
+                ],
+              },
+            ),
+        ),
+      )
+      .andWhere("application.location.id = :locationId", { locationId })
+      .andWhere("application.applicationStatus = :applicationStatus", {
+        applicationStatus: ApplicationStatus.Completed,
+      })
+      .andWhere("application.isArchived = false");
+
+    if (paginationOptions.searchCriteria) {
+      applicationQuery
+        .andWhere(
+          new Brackets((qb) => {
+            qb.where(getUserFullNameLikeSearch()).orWhere(
+              "application.applicationNumber Ilike :searchCriteria",
+            );
+          }),
+        )
+        .setParameter(
+          "searchCriteria",
+          `%${paginationOptions.searchCriteria.trim()}%`,
+        );
+    }
+
+    applicationQuery
+      .orderBy(
+        this.transformToEntitySortField(
+          paginationOptions.sortField,
+          paginationOptions.sortOrder,
+        ),
+      )
+      .offset(paginationOptions.page * paginationOptions.pageLimit)
```

Comment:
Can we use `skip` instead of `offset` here as `offset` may not work correctly in case of joins.

Reply:
what is the issue you are pointing here @sh16011993? Actually, we here using offset as there was some issue with the skip previously, I don't quite recall it.  @andrewsignori-aot @dheepak-aot @guru-aot @andrepestana-aot do you guys recall it

Reply:
I was just trying to suggest the below:

![image](https://github.com/bcgov/SIMS/assets/7859295/e7b79ad8-e832-4253-8c8a-9161792991cb)

I also found an additional link for the same:
https://www.kindacode.com/snippet/pagination-in-typeorm-find-options-querybuilder/#Using_QueryBuilder

Reply:
The last time that we went through this the conclusion was:
`take` + `skip` will give the precise results but they are not executed on DB, meaning, typeorm will not add anything extra to the query and the pagination will happen in memory, basically killing the idea of why we do pagination.
`limit` and `offset` will actually add extra instructions to the query execution of the pagination on DB. The possible incorrect results are probably because if the query results in relationships where a `one to many` will result in `many`, the count can be affected. So, whether it's a problem or not depends on the query that we are executing.

I am not sure if the behavior state above changed or not. The basic way to check it is by updating the query to use `take` + `skip` and checking if the generated SQL changes.

Reply:
@andrewsignori-aot I understood the part that if the pagination happens in memory for `take` + `skip` , the purpose is defeated since all records are returned from the db and the filtering happens in memory. If the `offset`+ `limit` returns the records correctly, then we can keep it as is.